src/compiler/utilities.ts

```diff
@@ -1471,14 +1471,75 @@ namespace ts {
         return getSourceTextOfNodeFromSourceFile(sourceFile, str).charCodeAt(0) === CharacterCodes.doubleQuote;
     }
 
+    export function getJSInitializerSymbol(symbol: Symbol) {
+        if (!symbol || !symbol.valueDeclaration) {
+            return symbol;
+        }
+        const declaration = symbol.valueDeclaration;
+        const e = getDeclaredJavascriptInitializer(declaration) || getAssignedJavascriptInitializer(declaration);
+        return e ? e.symbol : symbol;
+    }
+
+    export function getDeclaredJavascriptInitializer(node: Node) {
+        if (node && isVariableDeclaration(node) && node.initializer) {
+            return getJavascriptInitializer(node.initializer) ||
+                isIdentifier(node.name) && getDefaultedJavascriptInitializer(node.name, node.initializer);
+        }
+    }
+
+    export function getAssignedJavascriptInitializer(node: Node) {
+        return node &&
+            (isBinaryExpression(node) && node.operatorToken.kind === SyntaxKind.BarBarToken || isPropertyAccessExpression(node)) &&
+            node.parent && isBinaryExpression(node.parent) &&
+            node.parent.operatorToken.kind === SyntaxKind.EqualsToken &&
+            (getJavascriptInitializer(node.parent.right) || getDefaultedJavascriptInitializer(node.parent.left as EntityNameExpression, node.parent.right));
+    }
+
+    export function getJavascriptInitializer(initializer: Expression) {
+        if (isCallExpression(initializer)) {
+            const e = skipParentheses(initializer.expression);
+            return e.kind === SyntaxKind.FunctionExpression || e.kind === SyntaxKind.ArrowFunction ? initializer : undefined;
+        }
+        if (initializer.kind === SyntaxKind.FunctionExpression || initializer.kind === SyntaxKind.ClassExpression) {
+            return initializer;
+        }
+        if (isObjectLiteralExpression(initializer) &&
+            (initializer.properties.length === 0 ||
+             isBinaryExpression(initializer.parent) && isPropertyAccessExpression(initializer.parent.left) && initializer.parent.left.name.escapedText === "prototype")) {

```

Comment:
kinda weird to be checking up the parent chain here.. can we do it higher up instead, e.g. getDeclaredJavascriptInitializer